pkg/secretless/plugin/sharedobj/available_plugins_test.go

```diff
@@ -74,63 +75,144 @@ func TestAllAvailablePlugins(t *testing.T) {
-			pluginDir string,
-			checksumfile string,
-			logger log_api.Logger,
-		) (plugin.AvailablePlugins, error) {
-			return &allExternalPlugins, nil
-		}
-
-		// Create the test subject
-
-		allPlugins, err := AllAvailablePluginsWithOptions(
-			"",
-			"",
-			mock.GetInternalPlugins,
-			getExternalPlugins,
-			mock.NewLogger(),
-		)
-
-		// Test
-
-		assert.Nil(t, err)
-		if err != nil {
-			return
-		}
-
-
-		// Define our expected results (without override)
-		expectedHTTP := mock.AllHTTPPlugins()
-		expectedTCP := mock.AllTCPPlugins()
-
-		// Sanity check: Not equal without the override
-		assert.NotEqual(t, expectedHTTP, allPlugins.HTTPPlugins())
-		assert.NotEqual(t, expectedTCP, allPlugins.TCPPlugins())
-
-		// Now do the override
-		expectedHTTP["intHTTP2"] = httpOverride
-		expectedTCP["intTCP1"] = tcpOverride
-
-		// Now it should be equal
-		assert.Equal(t, expectedHTTP, allPlugins.HTTPPlugins(), "HTTP override failed")
-		assert.Equal(t, expectedTCP, allPlugins.TCPPlugins(), "TCP override failed")
-	})
+	// Test detection of conflicting external plugin names.
+	httpOrTCPNameConflict := "conflicts with same-named"
+	TestCases := []struct {
+		description       string
+		addPluginType     string // "internal" or "external"
+		addPluginProtocol string // "HTTP" or "TCP"
+		addPluginName     string
+		expectedPanic     string
+	}{
+		{
+			description: "There are no panics without duplicate plugin names",
+		},
+		{
+			description:       "An internal TCP plugin and internal HTTP plugin are loaded with same name",
+			addPluginType:     "internal",
+			addPluginProtocol: "TCP",
+			addPluginName:     "intHTTP2",
+			expectedPanic:     httpOrTCPNameConflict,
+		},
+		{
+			description:       "An external HTTP plugin and internal HTTP plugin are loaded with same name",
+			addPluginType:     "external",
+			addPluginProtocol: "HTTP",
+			addPluginName:     "intHTTP1",
+			expectedPanic:     httpNameConflict,
+		},
+		{
+			description:       "An external HTTP plugin and internal TCP plugin are loaded with same name",
+			addPluginType:     "external",
+			addPluginProtocol: "HTTP",
+			addPluginName:     "intTCP2",
+			expectedPanic:     tcpNameConflict,
+		},
+		{
+			description:       "An external HTTP plugin and external TCP plugin are loaded with same name",
+			addPluginType:     "external",
+			addPluginProtocol: "HTTP",
+			addPluginName:     "extTCP1",
+			expectedPanic:     httpOrTCPNameConflict,
+		},
+		{
+			description:       "An external TCP plugin and an internal HTTP plugin is loaded with the same name",
+			addPluginType:     "external",
+			addPluginProtocol: "TCP",
+			addPluginName:     "intHTTP1",
+			expectedPanic:     httpNameConflict,
+		},
+		{
+			description:       "An external TCP plugin and an internal TCP plugin is loaded with the same name",
+			addPluginType:     "external",
+			addPluginProtocol: "TCP",
+			addPluginName:     "intTCP1",
+			expectedPanic:     tcpNameConflict,
+		},
```

Comment:
I think this is still missing
- two external HTTP
- two external TCP
from the test cases

Reply:
Per our phone conversation, I'll move the uniqueness checks to happen earlier in the plugin discovery process, and retrofit/move the UT to apply to the place where those uniqueness checks land. Thanks for walking through this!